R/backport_linter.R

```diff
@@ -0,0 +1,83 @@
+#' @describeIn linters that checks for usage of unavailable functions. Not reliable for testing r-devel dependencies.
+#' @export
+backport_linter <- function(r_version = "4.0.3") {
+  function(source_file) {
+    if (is.null(source_file$xml_parsed_content) || r_version >= "r-devel") return(list())
+
+    xml <- source_file$xml_parsed_content
+
+    #browser()
+
+    names_xpath <- "//*[self::SYMBOL or self::SYMBOL_FUNCTION_CALL]"
+    all_names_nodes <- xml2::xml_find_all(xml, names_xpath)
+    all_names <- xml2::xml_text(all_names_nodes)
+
+    # guaranteed to include 1 by early return above; which.min fails if all TRUE (handled by nomatch)
+    needs_backport_names <- backports[1:(match(FALSE, r_version < names(backports), nomatch = length(backports)) - 1L)]
+
+    # not sapply, which may over-simplify -- cbind makes sure apply works
+    needs_backport <- do.call(cbind, lapply(needs_backport_names, function(nm) all_names %in% nm))
```

Comment:
I think there's a tricky part where many developers do their own backports in their code if necessary. For example, they define a `isFALSE <- function(x) identical(x, FALSE)` in the package first and then use it in other places (I don't see this function in the `backports` list?)

In this case, the approach seems to always give false results?

Reply:
good catch! I missed it because it's tucked at the bottom of another bullet point. this would have been a lot easier if the formatting of `NEWS` was more consistent :sweat_smile:

Reply:
see above for catching custom defines.

Reply:
Curious if we should allow explicit whitelisting. A user with a `backport` lint could take appropriate measures (e.g. import {backports}) and then edit the `.lintr` file like `backports_linter(safe = "fun_that_has_a_local_fallback_implementation")`.
WDYT?

Reply:
that's what I'm thinking yep, same idea as my comment above about `skip=`.